mqtt.go

```diff
@@ -19,598 +24,613 @@ const (
+	// we don't put these in the mqttConn's constructor.
+	client.wgSend.Add(1)
+	go conn.runPacketWriter(&client.wgSend)
+	client.wgRecv.Add(1)
+	go conn.runPacketReader(&client.wgRecv)
 
-func (mc *mqttConn) getPacketID() uint16 {
-	mc.packetID += 1
-	if mc.packetID == 0 {
-		mc.packetID = 1
-	}
-	return mc.packetID
+	// Connect and Subscribe here. These are exposed for now, but we might
+	// want to move them to non-public
+	return client
 }
 
-func (mc *mqttConn) connect() error {
-	logInfo("[MQTT] doing CONNECT handshake...")
+// Const function, but we don't want to pass client by value
+func (client *MqttClient) IsConnected() bool {
+	return client.isConnected
+}
 
+// exported for now, but pretty sure this needs to be called only from
+// NewMqttclient. This blocks until we get a Connection Ack
+func (client *MqttClient) Connect() error {
+	user, pwd := client.delegate.AuthInfo()
 	p := packet.NewConnectPacket()
 	p.Version = packet.Version311
-	p.Username = strconv.FormatUint(mc.dc.DeviceID, 10)
-	if !mc.dc.TLSClientAuth {
-		p.Password = mc.dc.AuthToken
-	}
+	p.Username = user
+	p.Password = pwd
 	p.CleanSession = true
 
-	if err := mc.sendPacket(p); err != nil {
+	respChan, err := client.conn.SendPacket(p)
+	if err != nil {
+		logError("[MQTT] failed to send packet: %s", err.Error())
 		return err
 	}
 
-	r, err := mc.stream.Read()
-	if err != nil {
-		logError("[MQTT] failed to read from stream: %s", err.Error())
+	var pkt packet.Packet
+	client.wgRecv.Add(1)
+	if pkt, err = client.receivePacket(respChan); err != nil {
+		logError("[MQTT] failed to receive packet %s", err)
 		return err
 	}
 
-	if r.Type() != packet.CONNACK {
-		logError("[MQTT] received unexpected packet %s", r.Type())
+	if pkt.Type() != packet.CONNACK {
+		logError("[MQTT] received unexpected packet %s", pkt.Type())
 		return errors.New("unexpected response")
 	}
 
-	ack := r.(*packet.ConnackPacket)
+	ack := pkt.(*packet.ConnackPacket)
 	if ack.ReturnCode != packet.ConnectionAccepted {
 		return fmt.Errorf("connection failed: %s", ack.ReturnCode.Error())
 	}
 
+	// If we made it here, we consider ourselves connected
+	client.isConnected = true
+
+	return nil
+}
+
+// This blocks until we receive an ACK (really, an EOF) from the server and
+// all our connections are done
+func (client *MqttClient) Disconnect() error {
+	defer func() {
+		client.delegate.Close()
+		client.wgSend.Wait() // wait for packet sender to finish
+
+		// Disconnects don't send a packet. Just make sure we wait for the EOF
+		defer client.wgRecv.Wait()
+	}()
+
+	// Creates a new context, using the client's timeout
+	p := packet.NewDisconnectPacket()
+	_, err := client.conn.SendPacket(p)
+	if err != nil {
+		logError("[MQTT] failed to send packet: %s", err.Error())
+		return err
+	}
+	client.conn.Close()
+
 	return nil
 }
 
-func (mc *mqttConn) subscribe(subs []mqttSubscription) error {
+// Add the subscriptions from the delegate
+func (client *MqttClient) Subscribe() error {
```

Comment:
We can not change the subscriptions with this implementation.
It may not be a problem for now but it will be better if we can change the subscription.

Reply:
For Mode use, we only have fixed topics, but it is not a general solution. I'll think about it more.